File: Programmers_Python/lv2_방문길이.py
def solution(dirs):
    dxy = {
        'U': (1, 0),
        'D': (-1, 0),
        'L': (0, -1),
        'R': (0, 1)
    }
    y, x, visited = 5, 5, set()

    for d in dirs:
        dy, dx = dxy[d]
        ny, nx = y+dy, x+dx
        if not (0 <= ny <= 10 and 0 <= nx <= 10):
            continue
        visited.add((y,x,ny,nx))
        visited.add((ny,nx,y,x))
        y, x = ny, nx
    return len(visited)//2


if __name__ == '__main__':
    dirs = "LULLLLLLU"
    print(solution(dirs))


# 처음엔 무작정 bfs로 풀려고했는데 bfs가 아니였음
# 어떤 점을 단순히 True,False처리하는 것도 어떤 방향에서 오느냐에 따라 다르기때문에 그럼 안됨
# 시작,끝 방향을 그래프 문제 풀때처럼 (시작, 끝), (끝, 시작) 두번씩 set에 넣어주어 처음간 길 카운트

File: Programmers_Java/Programmers_Java/src/lv2_방문길이.java
import java.util.*;

public class lv2_방문길이 {
    static HashMap<Character, int[]> cmd = new HashMap<>();
    static Set<List<Integer>> visited = new HashSet<>();
    public static int solution(String dirs) {
        cmd.put('U', new int[]{-1, 0});
        cmd.put('D', new int[]{1, 0});
        cmd.put('R', new int[]{0, 1});
        cmd.put('L', new int[]{0, -1});

        int curX = 5, curY = 5;
        for (char c : dirs.toCharArray()) {
            int[] d = cmd.get(c);
            int nx = curX+d[0];
            int ny = curY+d[1];
            if (!(0<=nx && nx <=10 && 0<=ny && ny<=10)) continue;
            visited.add(Arrays.asList(curX, curY, nx, ny));
            visited.add(Arrays.asList(nx, ny, curX, curY));
            curX = nx;
            curY = ny;
        }
        return visited.size()/2;
    }

    public static void main(String[] args) {
        String dirs = "LULLLLLLU";
        System.out.println(solution(dirs));
    }
}

/*
* set에 원시타입 배열이 아닌, Integer로 사용하기
* */